File: SimSim/Entities/Simulator.m
//
//  Simulator.m
//  SimSim
//
//  Created by Artem Kirienko on 29.09.2017.
//  Copyright © 2017 DaniilSmelov. All rights reserved.
//

#import "Simulator.h"

//============================================================================
@interface Simulator ()

@property (nonatomic, strong) NSDictionary* properties;
@property (nonatomic, strong) NSString* path;
@property (nonatomic, strong) NSDate* date;

@end


//============================================================================
@implementation Simulator

//----------------------------------------------------------------------------
+ (instancetype)simulatorWithDictionary:(NSDictionary*)dictionary path:(NSString*)path
{
    return [(Simulator*)[[self class] alloc] initWithDictionary:dictionary path:path];
}

//----------------------------------------------------------------------------
- (instancetype) initWithDictionary:(NSDictionary*)dictionary path:(NSString*)path
{
    self = [super init];
    
    if (self)
    {
        self.properties = dictionary;
        self.path = path;
        
        NSError* error = nil;
        NSDictionary * attrs = [[NSFileManager defaultManager] attributesOfItemAtPath:path error:&error];
        if (attrs && !error)
        {
            self.date = [attrs fileModificationDate];
        }
    }
    
    return self;
}

//----------------------------------------------------------------------------
- (NSString*) name
{
    return self.properties[@"name"];
}

//----------------------------------------------------------------------------
- (NSString*)os
{
    NSString* runtime = [self.properties[@"runtime"] stringByReplacingOccurrencesOfString:@"com.apple.CoreSimulator.SimRuntime." withString:@""];
    runtime = [runtime stringByReplacingOccurrencesOfString:@"OS-" withString:@"OS "];
    runtime = [runtime stringByReplacingOccurrencesOfString:@"-" withString:@"."];
    return runtime;
}

@end

File: SimSim/Entities/Simulator.swift
//
//  Simulator.swift
//  SimSim
//

import Foundation

//============================================================================
struct Simulator
{
    var path: String
    var date: Date
    private let properties: [AnyHashable : Any]
    
    //----------------------------------------------------------------------------
    init(dictionary: [AnyHashable : Any], path: String)
    {
        self.properties = dictionary
        self.path = path
        
        if let attrs = try? FileManager.default.attributesOfItem(atPath: path),
           let modificationDate = attrs[FileAttributeKey.modificationDate] as? Date
        {
            self.date = modificationDate
        }
        else
        {
            self.date = Date()
        }
    }
    
    //----------------------------------------------------------------------------
    var name: String
    {
        return properties["name"] as? String ?? "Unknown Simulator"
    }
    
    //----------------------------------------------------------------------------
    var os: String
    {
        guard let runtime = properties["runtime"] as? String else
        {
            return "Unknown OS"
        }
        
        return runtime.replacingOccurrences(of: "com.apple.CoreSimulator.SimRuntime.", with: "")
            .replacingOccurrences(of: "OS-", with: "OS ")
            .replacingOccurrences(of: "-", with: ".")
    }
    
    //----------------------------------------------------------------------------
    func pathForAppGroup(withUUID uuid: String) -> String
    {
        return path + "data/Containers/Shared/AppGroup/\(uuid)/"
    }
}
